### ISSUE TITLE
Annotations on typedef-ed types

### ISSUE THREAD

**Issue by vgurevich** [p4lang/p4-spec#815]
There are a number of types processed by P4 programs that traditionally use specialized literals. Examples are Ethernet addresses (`00:11:22:33:44:55`), IPv4 Addresses (`192.168.1.1`), IPv6 addresses (fe80::dbb9:94b3:1a21:9b44) etc. From P4 perspective, however, these types are simple unsigned integers or bit strings (`bit<48>`, `bit<32>` and `bit<128>` correspondingly). 

These literals do not matter from P4 perspective, but it is more than nice if the control plane (e.g. a CLI) would be able to accept such literals on the input and (especially) automatically display values in a proper form on the output.

A related problem is the desired radix for the values. For example, it is customary to display VLAN IDs or counter values in decimal, while many other fields are more proper to be displayed in hex or binary.

Unfortunately, the most natural solution, such as an annotation on the type:
```
typedef bit<48> mac_addr_t @type("mac_addr");
typedef bit<32> ipv4_addr_t @type("ipv4_addr");
typedef bit<12> vlan_id_t @radix(10);
```
does **not** work due to the fact that typedef establishes **full** type equivalency and thus all annotations are lost.

`type` does  seem to preserve annotations, but doesn't work either, since the defined types are extremely restricted.

So, currently we are stuck annotating **each and every** field or variable of the given type, which is really bad. 

One proposal might be to change how typedef works internally. Another is to remove restrictions from `type` or allow the programmer to specify restrictions explicitly if they so desire.

**Comment by jafingerhut** [comment]
Linking to this related issue, which proposed the possibility of enabling many of the operations that are currently defined on bit<W> types, so that they would also work on values declared as type foo_t where foo_t is declared as `type bit<32> foo_t;`: https://github.com/p4lang/p4-spec/issues/616

Without any changes to the latest P4_16 language spec (v1.2.0, but also true with whatever version that `type` was introduced in), you can take values of type `foo_t`, cast them to `bit<32>`, do whatever operations you want on them as that type, then cast the result back to type `foo_t` if you want to assign it to a variable/field of type `foo_t`.  Annoying, sure, and not something you probably want to recommend.

Even with the inheriting of operations mentioned in the issue above, if you had a variable `foo_t x` that you wanted to increment, I think you would need to do this: `x = x + (foo_t) 1;`, unless one also made changes to allow automatic casting from constant insteger literals to types like `foo_t`.

**Comment by jafingerhut** [comment]
Also, a puzzle question for you, related to your questions about possibly changing how `typedef` works.

Suppose you have these typedef's:
```
typedef bit<32> mytype1_t;
typedef bit<32> ipv4_addr_t;
```
Add whatever different annotations you want on mytype1_t and ipv4_addr_t to be as different from each other as you want.

Now declare a couple of variables:
```
mytype1_t a;
ipv4_addr_t b;
```
Now suppose somewhere in the program you have the expression `(a+b)`.  I believe the language spec and the compiler need to be able to determine a type for that expression.  What type should it be?  If `mytype1_t`, why that one, and not `ipv4_addr_t`?

A nice property of changing those `typedef` to `type` is that even though you will need at least one case to do `(a+b)`, you and the compiler will know what the type of the resulting expression is.

**Comment by vgurevich** [comment]
@jafingerhut  -- thanks for the detailed responses!

I think that it is precisely for the reasons you mention that I'd like to have a `typedef` that preserves the additional info rather than `type`. 

As you can imagine, casting things back-and-forth is not very convenient and in my opinion does not add to the program readability. Neither it serves the modularity since the base type has to be known (not a biggie for ipv4_addr_t, but a bigger problem for other annotations I mentioned). Obviously, this can also be solved by having a special easy syntax for the casting to basetype, but it is still too much to type and see for no good reason. 

I think that the implicit casts, similar to what we introduced for serializable enums might be a better answer to the question you posed. If we assume their existence than the type of `(a+b)` in your question will be `bit<32>` with an optional warning from the compiler.  Similarly, the type of `(a+1)` will be `bit<32>`, but without a warning.

**Comment by mbudiu-vmw** [comment]
The problem is that declarations can have annotations, but not expressions. If something participates in an expression there are no rules on how to propagate annotations. If you add 1 to an IP address, is it still an IP address?

If you don't want arithmetic on these types then perhaps `type` will suffice.

Perhaps we can add something to allow users to define operations on values with `type` types. For example, you could define masking (&):

```
type bit<32> IPAddress;

IPAddress operator&(in IPAddress address, in bit<32> mask) {
    return (IPAddress)(bit<32>address & mask);
}
```

Then the compiler would substitute this function whenever you call `&` on expressions with suitable types.

**Comment by jafingerhut** [comment]
@vgurevich I will focus in this particular comment on the idea of automatic casting from a type like `type bit<32> ipv4_addr_t;` to `bit<32>`, and what I think that would mean with only that change.  I am assuming that there is no automatic casting in the other direction from `bit<32>` to `ivp4_addr_t`, just as there is not for serializable enums.  Not sure what you were thinking there, but you did say "similar to what we introduced for serializable enums", so I will start with something as similar as possible and see where that leads.

So with P4_16 v1.2.0, we have to do two casts in this example:
```
type bit<32> ipv4_addr_t;
ipv4_addr_t a1;

a1 = (ipv4_addr_t) (((bit<32>) a1) & 0xffff0000);
```
With a change that would allow automatic casting from `ipv4_addr_t` to `bit<32>`, but not in the other direction, we can remove one of the casts, like this:
```
a1 = (ipv4_addr_t) (a1 & 0xffff0000);
```
Good enough?  Or were you hoping the cast was also automatic in the other direction.

I am not sure, but I believe that any proposal that includes automatic casting in both directions changes `type` back into `typedef`, or at least if it does not make them equivalent, I do not see why not.

**Comment by vgurevich** [comment]
@mbudiu-vmw  -- let's concentrate on the use case. Perhaps, I should not have made any comments on the desired implementation (and I thought I carefully said that these are only some ways that I can see).

Ultimately, if I declare a type `ipv4_addr_t` and I have a lot of fields of that type, every time there is a table that matches on such a field or an action that uses an argument of that type, I want the control plane to be able to use 'aaa.bbb.ccc.ddd' notation, especially when displaying this field. 

I also want to make sure I am not restricted in what I can do with that field, compared to using bit<N>.

I also want to make sure that the code is readable and does not contain unnecessary casts (and especially multiple casts per expression). 

Is that an unreasonable requirement that we can address?

In your particular example, the result will not be visible to the control plane or it will be visible only through very specialized debugging tools, where precision might be more important than ease-fo-use.

**Comment by vgurevich** [comment]
@jafingerhut -- I would say that this is not a bad compromize at all, although going all the way would be even better :)

**Comment by jafingerhut** [comment]
@vgurevich I just created 4 very similar P4_16 programs in this public repository, named weird-case1.p4 through weird-case4.p4

https://github.com/jafingerhut/p4-guide/tree/master/p4runtime-translation

See the README there for some notes on these tiny programs, the small differences between them, and the questions raised at the end.  I don't have good answers for those questions.  Maybe you do?

**Comment by jafingerhut** [comment]
Since I wrote my previous comment, I obsessed a bit and came up with one possible proposal for answering those questions, that requires fewer casts for many expressions than P4_16 does today.  I have not thought long and hard about them yet, and they could contain multiple errors and/or bad ideas that a P4 compiler developer or other language design expert could spot.  It is at the end of the README linked in my previous comment.

**Comment by jafingerhut** [comment]
@stefanheule @konne-google There is no specific proposal on the table yet to change the P4_16 language spec regarding this, but the questions raised in this issue naturally lead to questions of whether enabling P4_16 developers to write programs using types defined as `type bit<32> IPv4Addr_t` with a `@p4runtime_translation` annotation on them, they should be able to more freely mingle values of type `IPv4Addr_t` and `bit<32>` with each other in arithmetic like `&`, `+`, etc.

I am pinging you here to ask you: Is your use of `type` in P4_16 as an opaque type, at least in part _because_ you want a P4 compile time error if you ever attempt to mingle such types, in any way, without an explicit cast?

If yes, then that could affect any proposal made related to this issue, e.g. perhaps a `type` declaration to work exactly like it is specified today, but some new option might be introduced that meant 'the type created using _this_ `type` declaration is allowed to mingle more freely with other values of type `bit<32>`'

**Comment by vgurevich** [comment]
@jafingerhut -- First of all, thank you so much for putting so much thought and effort into the problem. I am really glad you got obsessed and decided to revive the issue. 

Generally, I like the idea of the proposal, but before we dive deep into the details, let me offer a slightly different take on that, which might actually simplify things quite a bit. I did voice the same idea during the `type` discussion, but I think it got somewhat lost.

The idea is that all P4 types have two "faces",  that are relatively independent:
- the data plane "face"
- the control plane "face"

The problem is that P4 is not very good at articulating the existence of the second face and it tries to pretend that it doesn't exist, while in reality it does -- it is simply a set of conventions. Current conventions are like this:
- Data Plane: `bit<1>`  -- Control Plane: `uint8_t`
- Data Plane `bit<2>`  -- Control Plane: uint8_t`
- Data Plane: `bit<12>` -- Control Plane: `uint16_t`
- Data Plane: `bit<16>` -- Control Plane: `uint16_t`
- Data Plane: `bit<48>` -- Control Plane: `uint64_t` **or** `uint8_t[6]`
- etc.

You get the point -- data plane and control plane "faces" of the same P4 type are generally independent.

What I think we really need is the ability to have this distinction in the program, so that one can have multiple types that have the same data plane "face", but different control plane "faces" (we can discuss what that means) otherwise.

As we all know, `typedef` creates types with **both** the same data and control plane faces, while `type` forces us to have types with different data plane faces ("island of values" as you put it).  What is missing is a construct that creates equivalent data plane types with different control plane attributes.

Why I think this makes things easier? Because the underlying type (the data plane "face") will be the same and thus, there is no ambiguity when it comes to expressions. On the control plane side, using the default face (e.g. `uint32_t` for `bit<32>`) is always safe and acceptable. If someone wants to override it, it is also possible. 

I think it would pretty much summarize the complex rules. 

There are other advantages to this approach. For example, if you modify the program and change 15-bit index to 17-bits your control plane interface might change going from uint16_t to uint32_t. Now, it is possible to avoid that. 

What do you think?

**Comment by jafingerhut** [comment]
If you have the weird-case1.p4 program, and diff'd it against the weird-case4.p4 program, then the following is just a repeat of the question asked there.

Suppose two types have the same data plane face, but different control plane faces.  For the sake of concreteness here, those two data plane faces could be different P4_16 types declared using the `type` keyword, e.g. `type bit<32> dpFaceA;' and `type bit<32> dpFaceB;`, which could have completely different types in the control plane.  No problem so far.

Do you want to be able to do this in a P4_16 program?

```
    dpFaceA a;
    dpFaceB b;

    table t1 {
        key = {
            a + b : exact @name("myLookupKey");
        }
        // .. rest of table definition omitted
    }
```
What is the control plane face of this table key?

**Comment by vgurevich** [comment]
@jafingerhut -- yes, and the control plane face for `a+b` is whatever is the default for bit<32>, since that's the data plane face type. If someone wants to have a different control plane face, they can use a cast or an annotation.

**Comment by jafingerhut** [comment]
So you are saying, you would like it if an expression contains a mix of data plane values with different control plane types, then the control plane type should be quietly switched to `bit<W>` for whatever bit width W is the data plane width.

But if an expression only involves exactly one of those data plane types (however we define that, for all expressions that one can write in the language), then the result should have its special control plane face.

Casts can override the control plane face type of an expression.

Except for overriding the default control plane face type of an expression, casts to other types are only needed to change between signed/unsigned, or different bit widths.

Does that sound anywhere close to what you are thinking?

**Comment by jafingerhut** [comment]
Just to make the example just slightly longer, to see if I am understanding you:
```
    type bit<32> myTypeA;   // this has control plane type "dotted decimal IPv4 address string"
    type bit<32> myTypeB;   // this has the control plane type "hexadecimal string"

    myTypeA a;
    myTypeB b;
    bit<32> c;    // this has the control plane type "whatever the default is for bit<32>"

    table t1 {
        key = {
            a: exact;   // control plane type "dotted decimal IPv4 address string"
            b: exact;   // control plane type "hexadecimal string"
            c: exact;   // control plane type "whatever the default is for bit<32>"
            a+b: exact;  // control plane type "whatever the default is for bit<32>"
        }
        // ... rest of table definition omitted
    }
```
Do all of those control plane types match what you would expect?

**Comment by vgurevich** [comment]
@jafingerhut  -- yes, I think this is the easiest. We can certainly create more elaborate rules, but the important thing is that these rules will not affect the program semantics.

**Comment by jafingerhut** [comment]
If the control plane face is just a nicety for debug purposes, then yes, the control plane face doesn't matter that much.

But what if all of the control plane software is written to rely upon a particular control plane face being used?  Then it isn't just a convenience.  It is critical.

**Comment by vgurevich** [comment]
@jafingerhut  -- I think that explicitly stating that every P4 type has two faces is actually beneficial to the both the clarity and stability of the control plane. It doesn't take away anything that we have today, but (a) formalizes the relationship that exists whether we like it or not and (b) provides clear control over the control plane face.

In reality, I would expect that P4 Runtime or a similar API would be relatively agnostic to these annotations and continue to use integers in the actual messages. The translation will probably happen in the layers above (such as  higher-level API bindings or CLIs). 

As I mentioned, this can also allow some additional benefits. For example, today there are APIs where changing an action parameter (e.g. a nexthop index) from `bit<15>` to `bit<17>` results in a change from `uint16_t` to `uint32_t` on the control plane side with serious consequences.  A serious data plane designer can force `uint32_t` to always be the control plane face of the corresponding type (nexthop_id_t) thus providing better compatibility.

**Comment by jafingerhut** [comment]
@vgurevich FYI, folks at Google involved in the P4Runtime API working group have been pushing for using strings (OK, sequences of bytes, but much longer than the data plane sizes are) in the P4Runtime API messages for things that in the data plane, like nexthop_id, are `bit<10>` or `bit<12>` in size, using the P4Runtime translation feature.  They _want_ the P4Runtime API messages to be noticeably bigger than the corresponding data type sizes, so they can put human-readable strings into them.  Not for things that they would try to download 1 million of in 50 milliseconds, of course, but things where they seem to prefer debug ability of reading P4Runtime API messages over message size concerns.

I am not saying there is necessarily a wide groundswell of desire for such a capability, but at least some people seem to want it.

**Comment by jafingerhut** [comment]
@mbudiu-vmw @antoninbas A question on p4c front end passes and P4Runtime API generation.

Is it true that P4Runtime API generation is today done after the last front end pass, before any mid end passes?

When I use --top4 option to print out the P4_16 code after every pass, and look at programs that use `typedef` in the original source code, there are still `typedef` definitions, and declarations of variables/struct fields that use those types, preserved all the way until the last front end pass.

If that is all true, does that mean that the p4c IR data structures also preserve typedef types until then?

If so, it seems that it would also be potentially useful for achieving Vladimir's goals to support `@p4runtime_translation` annotations on `typedef` definitions.

As far as I can see, the only thing it would _not_ allow you to automatically infer would be arbitrary expressions used as table key fields.  We could have a simple rule like: "If the table key expression is just a variable name, or field name of a struct or header, use its typedef type.  If it is anything else, use bit<W>, unless they put an explicit cast on the entire expression, then use that cast type"

**Comment by mbudiu-vmw** [comment]
If what you want is for some objects to have two types then probably the right way to do this is to add the ability in P4 to declare objects that have separate data-plane and control-plane types. One way we could do it today is to add an annotation on a declaration `@controlPlaneType("bit<64>")` - unfortunately annotations cannot hold type names, only expressions, so it must be a string. But perhaps we can define a cleaner way.

If we need control plane types for expressions we need a type theory that explains how control plane types for expressions are inferred - the same way we have one for data plane types.

Each back-end can choose where to do P4 API generation; for p4test it's done right after the front-end: https://github.com/p4lang/p4c/blob/master/backends/p4test/p4test.cpp

**Comment by vgurevich** [comment]
@jafingerhut -- 

I think that Google's requirement can also be nicely supported by explicitly declaring control plane type for a given P4 type. Translation can as easily be done at the agent side too. 

It looks like that we reached the same conclusion, namely "it seems that it would also be potentially useful for achieving Vladimir's goals to support @p4runtime_translation annotations on typedef definitions.", which is pretty much the subject line of the issue (except that I didn't name a specific annotation). 

@mbudiu-vmw --

The first problem is to be able to do it on typedef-ed types, since the point is that the same data plane type might have different control plane faces.

The issue with expressions, etc. is not a problem, I believe as long as a particular, control-plane visible  expression has the default control plane type and can also be annotated separately.  I have an idea for a more elaborate theory, but I'd like start with this simple one first.

**Comment by jafingerhut** [comment]
Hmmm.  Thinking about this more, the only place using the P4Runtime API that an expression can appear are:

+ table search key fields.  In the test program switch.p4 in p4c, there are a little over 300 table search keys, and all but 9 of them are simply the name of a header or metadata field.  9 of them are "name mask 0x0ff000" (for some constant mask).
+ digest messages, which are most often explicitly declared structs in P4_16, which have explicitly declared types for every field.

So even though table search key fields can be arbitrary expressions, it is very unusual for them to be anything other than a single field, or myhdr.isValid(), which has a clear type of boolean (or bit<1>).

So the idea is definitely growing on me to really see whether this could be done in p4c -- add annotations to `typedef`, and if a table search key field is just a single field, then preserve its `typedef` name.

If for some reason local copy propagation can take assignments to variables and propagate those expressions into the table search key field, would p4c do that in a front end pass?  If it can in today's p4c implementation, would it be reasonable to disallow that behavior in the front end, before the control plane API is generated?

I am feeling hard pressed to see a reason why p4c could not do this.

**Comment by konne-google** [comment]
Thanks @jafingerhut for the ping! I believe that the current definition of `type` is exactly what we want, and we should *not* introduce implicit casts (in either direction).

The `type` definition provides the compiler/switch with a very powerful guarantee: the only operation performed on values of `type` is equality. This is very powerful knowledge because it allows the compiler/switch to remap values of `type` without breaking the program. e.g. we can have programs that represent ports as a type of bit width 32, or even as strings (that are much more readable than numbers), and the compiler/switch can then remap these values to more efficient representations (e.g. a hardware platform that only supports 128 ports could remap to a 7 bit int). 

Introducing automatic casts (even in just one direction) breaks this guarantee, and therefore I feel very strongly that we should not do it. Consider for example the port definition in PSA on a hardware platform that supports upto 128 ports. To the controller, port ids will be 32 bits, but because we have a `type` definition, the compiler/switch can actually represent these with just `7` bits.

```
@p4runtime_translation("p4.org/psa/v1/PortId_t", 32)
type bit<7> PortId_t;
```

With implicit casts, we will loose that ability. Consider the following P4 program, where the `acl` table tries to match on a set of ports. E.g. an `acl` rule could try to match on ports `0`, `3`, `4`, and `7` by installing a table entry with value `0b10011001` and mask `0b10011001`.

```
// each bit is one port
bits<128> ports = 1 << psa_metadata. ingress_port
table acl {
  key {
    ports : ternary
  }
}
```

If the compiler would now try to replace ports, everything would break. The above example rule wouldn't match on ports `0`, `3`, `4`, and `7` anymore, but some other unspecified set of ports that depends on the mapping. Therefore, it's very important that the compiler warns users of this problem, especially new users who aren't completely familiar with the port translation mechanism yet, and can easily end up in a very confusing situation here.

Again, I feel very strongly that we should keep the current behavior of `type`. If a less restrictive construct is needed, don't use `type` or use explicit casts which make it clear that you are on thin ice.

CC:
@stefanheule

**Comment by vgurevich** [comment]
Thanks, @jafingerhut !

While I am a little bit afraid to complicate the discussion, here are some relatively simple rules that should take care of 99% of all the cases in a very reasonable manner.

Here is the base example:
```
typedef bit<32> ipv4_addr_t @cp_type("ipv4");
typedef bit<32> index_t @cp_type("decimal");

ipv4_addr_t a;
ipv4_addr_t b;
index_t c;
bit<32> d;
```

1. The type of `a+b`, `a&b`, `~a`, etc. is `ipv4_addr_t` or in other words the annotations stay
2. The type of `a[31:24]` is `bit<8>`, and thus is definitely not `ipv4_add_t`. Similarly, the type of `a++b` is `bit<64>`. In other words, if the base type of the result is different, all annotations are cleared (but can be added via a cast or explicitly)
3. The type of `a+c`, etc. is `bit<32>`. In other words,  if there are different annotations, they are all cleared and we are back to the base type
4. The type of `a+1`, `a+32w10` and `a+d` is `ipv4_addr_t`. In other words, if one operand has annotations and another one doesn't annotations are preserved.

I think these 4 rules will provide the most natural handling for most of the cases. Note, that all these rules are completely compatible with the existing `typedef` semantic.

**Comment by jafingerhut** [comment]
@vgurevich You have just written down a shorter set of rules than the ones in the README I linked above.  I know, I write too much to possibly be interesting to read sometimes.  It is a character flaw that I will probably never correct :-)

In any case, I am motivated to talk to Mihai personally at some point and figure out what is possible, and try to learn if there are other objections that I do not yet understand for having something like this.

**Comment by vgurevich** [comment]
@jafingerhut  -- I did read the `README` and I felt that what you were trying to express is probably quite similar, but establishing clear equivalence was, indeed difficult :)

If you are not afraid of COVID-19 and would like to meet for lunch (with Mihai), I'm game too.

**Comment by vgurevich** [comment]
@ konne-google, 

So far, the discussion is centered around `typedef`, not `type`, so if you prefer the restricted types, you should not be affected.

**Comment by konne-google** [comment]
@vgurevich thanks, I did read the entire conversation, but it did see you guys were initially proposing to use `type` for these annotations with automatic casts. If that's not the case anymore, then I'm happy :)

**Comment by antoninbas** [comment]
@jafingerhut This is an answer to your question in https://github.com/p4lang/p4-spec/issues/815#issuecomment-598033044, I didn't not read this thread in details.

It is true that typedefs are preserved at the type P4Info generation happens. This is an observation I made a while back as I was fixing a bug in the P4Info generation code (https://github.com/p4lang/p4c/blob/master/control-plane/typeSpecConverter.cpp#L131), and which contradicts what I originally assumed at the time we introduced `type`.

However, as far as I'm concerned, this is somewhat of an implementation detail at this time, and it would have been equally acceptable for the compiler to eliminate all typedefs early-on in the frontend. If @mbudiu-vmw can guarantee that this is not going to change in the future, then I am happy to discuss `@p4runtime_translation` support for typedefs as well. My personal take is that it is not such a good idea, unless someone can come-up with clear rules for type inference of expressions involving these typedefs. I see that @vgurevich took a stab at this here: https://github.com/p4lang/p4-spec/issues/815#issuecomment-598837632.

**Comment by jafingerhut** [comment]
@antoninbas Thanks for the reply, and I agree with the direction of your comments, i.e. that the p4c implementation is not the language spec.  Still, we are in the perfect place to affect the spec and/or implementation in ways that are harmonious with each other, if we can figure out how.

I am also all for clear rules, but can you think of any places where inference of expressions is important, other than table key expressions?  That is, any other place where these inference rules would affect generating the control plane API?  If it is only that, then at least in the common case, the rule that will be used most often is "a header or struct field, or local variable, is the typedef that is is declared as", because that is what almost all such expressions are in practice.

If an expression is on the right hand side of an assignment, and it has no reason to be involved in the control plane API (as most assignment expressions are not, typically), then there is no reason for the compiler to derive its control plane API type, is there?

**Comment by antoninbas** [comment]
@jafingerhut I think you are pretty much correct. With v1model, it is possible to create a digest instance with an extern function call like this one: `digest(1024, {fields})`, and let the compiler infer the type of `{fields}`. But I would say `digest` in v1model is broken, at least from a control-plane API perspective, and PSA took the correct approach by using an extern type instead.

One idea I had a while back is to introduce the **ability** (I say ability so that P4 stays backward-compatible) to specific explicit types for match fields, which could look like this:
```
    typedef bit<32> IpAddr_t;
    @p4runtime_translation("foo.org/myarch/v1/PortId_t", 32)
    typedef bit<9> PortId_t;

    table t1 {
        key = {
            IpAddr_t hdr.ipv4.srcAddr & 0xffffff00: exact;
            PortId_t stdmeta.ingress_port: exact;
            bit<32> meta.whatever: ternary;
        }
        actions = { ... }
    }
```
or maybe
```
    typedef bit<32> IpAddr_t;
    @p4runtime_translation("foo.org/myarch/v1/PortId_t", 32)
    typedef bit<9> PortId_t;

    table t1 {
        key = {
            hdr.ipv4.srcAddr & 0xffffff00: exact -> IpAddr_t;
            stdmeta.ingress_port: exact -> PortId_t;
            meta.whatever: ternary -> bit<32>;
        }
        actions = { ... }
    }
```
or we could even grab this opportunity to provide the ability to have better control over match field names, without annotations:
```
    typedef bit<32> IpAddr_t;
    @p4runtime_translation("foo.org/myarch/v1/PortId_t", 32)
    typedef bit<9> PortId_t;

    table t1 {
        key = {
            hdr.ipv4.srcAddr & 0xffffff00: exact -> IpAddr_t srcAddrPrefix ;
            stdmeta.ingress_port: exact -> PortId_t port;
            meta.whatever: ternary -> bit<32> what;
        }
        actions = { ... }
    }
```
I'm sure people could come up with a better syntax. I don't know if there would be any drawback to this (besides the increased verbosity) or if there would be resistance to introducing something like that.

**Comment by jafingerhut** [comment]
One can already today use casts on table key field expressions, e.g.:
```
table t1 {
        key = {
            (IpAddr_t) (hdr.ipv4.srcAddr & 0xffffff00): exact;
            (PortId_t) stdmeta.ingress_port: exact;
            (bit<32>) meta.whatever: ternary;
        }
        actions = { ... }
    }
```
However, the second two would be redundant if stomata.ingress_port was type `PortId_t`, and meta.whatever was type `bit<32>`, and the first one would be redundant with some of what sound like fairly simple rules that Vladimir and I have been describing for the first table key, too (i.e. any expression that includes only variables/fields of type `typedef bit<32> IpAddr_t` and other things that are `bit<32>`, is `bit<32>`, unless you give an explicit cast that it is some other type).

**Comment by mbudiu-vmw** [comment]
key fields can have annotations:
```
key = {
            hdr.ipv4.srcAddr & 0xffffff00: exact @cp("IpAddr_t srcAddrPrefix");
            stdmeta.ingress_port: exact @cp("PortId_t port");
            meta.whatever: ternary @cp("bit<32> what");
        }
```

**Comment by mbudiu-vmw** [comment]
But I am not opposed to adding additional syntax for key fields if that solves most of the requirements.

**Comment by jafingerhut** [comment]
In case it isn't clear, I think the reason that Vladimir and I are being so persistent on the `typedef` approach is that a single annotation of the right kind on one `typedef` declaration, means that there is no need for N annotations/declarations/casts/whatever in N places where values of that type are used all over the code.

I believe the predominantly common case will be that someone will want values of a particular type to be the same control plane type wherever it is used throughout the program.  Just like the current `@p4runtime_translation` annotation on `type` declarations can, except without needing to cast so much when you _do_ want to do arithmetic on `typedef` type values (which that arithmetic already works in P4_16 today with no spec changes).

**Comment by mbudiu-vmw** [comment]
Antonin's proposal does more than change the type, it can also change the name.

**Comment by mbudiu-vmw** [comment]
We should perhaps add a new kind of type declaration that looks like typedef but has two types, one for the dataplane and one for the control plane.

**Comment by jafingerhut** [comment]
p4c today already allows `@name` annotations on every table key field.  It _requires_ them if the table key field is any expression more complex than a single field or variable name.  We can overhaul things with a second way to assign a name if you want, but I do not see a motivation for doing so (yet).

"We should perhaps add a new kind of type declaration ..."

```
@p4runtime_translation("my company.com/p4/...", 32)
//.                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ control plane type
typedef bit<16> mytype_t;
//           ^^^^^^^^ data plane type
```
is at least one possibility, and already exists and works for `type` declarations.

**Comment by mbudiu-vmw** [comment]
that annotation is a bit ugly and the string makes it even worse.
If the type was explicit then the typechecker could check the control-plane types too.

**Comment by jafingerhut** [comment]
@vgurevich Are you expecting a P4 compiler to ever "check control-plane types"?  If so, can you give an example of what that means to you?

I have no idea what it would mean for a P4 compiler to "check the control-plane types", nor why that would be desirable.

Everything we have done with P4Runtime translation support so far is for the P4 program developer to simply say what the control-plane type is, and use it in the control plane API generation.

**Comment by mbudiu-vmw** [comment]
This should be judged from the point of view of the P4 programmer; we should provide the most natural way to do the tasks they need to do without compromising the integrity of the design.

**Comment by jafingerhut** [comment]
One property of having some kind of string there in the P4 source code is that it is _not_ limited to types that can be specified in P4_16 source code today, and probably never will be possible.

For example a control plane API type might be "dotted decimal strings with 3 dots in them, and each decimal number in the range [0, 255]", which I don't ever expect to be a P4_16 type.  It isn't even a type in any general purpose programming language I have ever heard of.

I am not saying that nothing better can be designed -- I am cautioning that if you are thinking control plane types should be limited to what P4_16 can specify, that seems way too limiting.

**Comment by mbudiu-vmw** [comment]
In this case we cannot do better than strings.

**Comment by vgurevich** [comment]
Oh, boy -- 2 hours offline and suddenly there is such a lively discussion! :)

Let me catch up and respond.

**Comment by vgurevich** [comment]
@jafingerhut  -- thank you for being so persistent and on the same "wavelength". 
@antoninbas  -- while I certainly like the idea of assigning table key fields names using language syntax instead of `@name()` annotation, I suggest we do it separately
@mbudiu-vmw -- the idea to declare that each type has two "faces" at the language level is certainly enticing and we should probably mention that somewhere, but I am afraid that this will, indeed, make things less flexible.

Details....

Totally agree with @jafingerhut that potential annotations are a property of a particular control plane interface and thus should not be limited. Thus, for now strings are probably better. 

Perhaps, we can have yet another `typedef` syntax that will accept these annotations as parameters, e.g:
```
typedef bit<32>("ipv4_addr") ipv4_addr_t;
typedef bit<32>("%o") octal_number;
```
The good news is that the algorithm suggested above only needs to check for presence and equality without trying infer any semantics from the strings. Obviously, more complex string names, suggested by Andy are also possible, although I like things short and sweet.

If we allow this, then `@cp_type()` annotations will probably used only very sparingly. If that doesn't work, I'll be happy with just an annotation (even though I try to avoid them as much as I can).

@antoninbas -- a while ago @smolkaj proposed the following syntax that I really liked:
```
table t {
   key = {
       src = min(h.ipv4.dst_addr, h.ipv4.src_addr) : ternary;
       dst = max(h.ipv4.dst_addr, h.ipv4.src_addr): ternary;
   }
   . . . 
}
```

**Comment by smolkaj** [comment]
(@vgurevich: A while ago, in 2016 :smile:)

**Comment by vgurevich** [comment]
Any thoughts on how can we move forward?

**Comment by jafingerhut** [comment]
Since early 2019, the P4Runtime API specification has documented the `@p4runtime_translation` annotation as a way to do this.  See Section 8.5.6 "User-defined types" here: https://p4.org/p4runtime/spec/master/P4Runtime-Spec.html#sec-user-defined-types

This has been implemented since early 2019 in p4c, with the following as a fairly extensive program to test this feature with many combinations of `type` and `typedef` definitions in a program, merged into p4c in May 2019: https://github.com/p4lang/p4c/blob/8dc6f510fada60c0ebc50937280484be1a5b7382/testdata/p4_16_samples/v1model-p4runtime-most-types1.p4

That test program was written at a time when I thought that supporting such an annotation was only useful possible for `type` declarations, although with this issue and my recent looking at intermediate p4c pass results, it seems that `typedef` could probably support such annotations, too.

One way forward, that is as similar to what was done in the past as possible, is to add support for this same `@p4runtime_translation` annotation, except enable its use on `typedef` declarations, too.

Or, we can design, document, implement, and write test cases for something new.

**Comment by vgurevich** [comment]
The main issue here is the (in)ability to attach **any** annotation to `typedef`-ed types, be it `@p4runtime_translation` or any other. This is what I think we were discussing.

Aside from that, P4Runtime is not the only API that people are interested in :) , so it is better to have a general mechanism that allows any annotation.

**Comment by jafingerhut** [comment]
So more generally, one way forward is to investigate in more detail whether what you would like to work, could actually work by using annotations `typdef`declarations, or whether there are issues with that approach that haven't been discovered yet.

**Comment by vgurevich** [comment]
We do know that they are not preserved/carried through

**Comment by jafingerhut** [comment]
@vgurevich @mbudiu-vmw @antoninbas

Vladimir, Mihai, and I have been discussing this for a while in other channels, and came to these conclusions, or in a few cases, proposals for a few more experiments and code investigation.  I am pinging Antonin Bas, who was not involved in these discussions, to ask if he minds skimming this over to see if I'm saying anything that he knows to be too far from the truth.

The current open source p4c compiler does preserve both type and typedef declarations in its intermediate representation all the way through the front end passes, and it replaces type declarations with typedef declarations fairly late in the mid end passes.  There is nothing fundamental in the language specification that requires this, nor is there anything that disallows it.  It is a property of the current implementation, and one we see no reason to change.

It thus seems like a fairly straightforward thing to extend the P4Runtime API generation code that exists in open source p4c, the part that handles the `@p4runtime_translation` annotation on `type` declarations, so that it also works for the same (or other) annotations on `typedef` declarations, at least for these kinds of things in the control plane API, where there are no expressions involved:

+ action parameters
+ table keys that are simply the name of a header field, with no operations involved, not even a bit slice.  These are by far the most common case for table key expressions.  The next most common, but probably under 10% of all cases seen in practice, are bit slice expressions, simple mask expressions like "field_name & 0x0f0f", and "hdr.hdr_name.isValid()" expressions.  Those are discussed further below.
+ lists expressions in digest calls, which are often simply lists of field names, so simple expressions like those in table expressions
+ TBD on lists expressions to hash function calls, but those are likely often substantially similar to the digest call case.

For the common cases, it is straightforward to look up the `typedef` name of any identifier in the p4c's IR, and to check whether it has any annotations.  This is roughly how the P4Runtime API generation looks up the `type` name of identifiers today, and then checks if there are any `@p4runtime_translation` annotations on them.

For the less common cases, like the table key expressions mentioned above, it seems that a reasonable approach is to say that when performing control plane API generation, there should not be any fancy/optimizing compiler transformation going on before that step, and perhaps even to describe them in the spec (a rough starting point of such a list is given below).  Thus the compiler's IR will be "reasonably close" to representing the source code, and P4Runtime API translation code (or other control plane API generation code) can be written that examines a table key expression like "hdr.ethernet.srcAddr & 0xff0000", in the context of these declarations:
```
@p4runtime_translation("p4.org/psa/v1/EthernetAddress", 48)  // or non-P4runtime API-specific annotations here instead of this, or in addition to it
typedef bit<48> Eth_t;

Eth_t srcAddr;
```
and performs whatever rules it wishes to infer a control plane API type, e.g. "any expression that is a field with a typedef type, on which any arithmetic operations are done with compile-time constant numbers, that leaves the result the same bit width, will have a control plane API type the same as the header field by itself".  That code would not be part of the P4_16 language spec, nor even the core of the compiler.  It would be specific to the control plane API generation part of the compiler.  (There might be multiple control plane API generation implementations that wish to use common code here -- the P4Runtime API generation code would be an open source example that others could base theirs on, if they wished).

The things that seem most likely additions to the P4_16 language specification might be the following:

+ The compiler should pay attention to and maintain annotations on both `type` and `typedef` declarations (perhaps already implied or explicitly stated in current spec -- I have not checked yet).

The kinds of things that should be done before control plane API generation in a compiler:

+ parsing
+ syntax checks
+ type checking
+ instantiation of all packages, control, parsers, and extern objects, from the `main` object "down", as described in 17.2 "Compile-time Evaluation" of the current language spec

The kinds of things that should not be done before control plane API generation:

+ any kinds of optimizations of expressions, e.g. common sub-expression elimination, local copy propagation, strength reduction, etc.  I think this could only serve to make the expressions used to generate the control plane API unnecessarily further from the source code.

It isn't clear to me whether dead code detection/elimination should or should not be done before control plane API generation.  It seems perfectly reasonable to me if it is not.  If someone wants a table not to be in the 'object code' of a target, they can delete it, or comment/ifdef it out.

**Comment by vgurevich** [comment]
Thank you, @jafingerhut for the write-up.

Spec-wise, I think it will be best if we explicitly codify the idea of the types having separate data plane and control plane "faces" and then further clarifying that everything that has been written about types, casting, etc. applies to the data plane "face" of the type, whereas the control plane "face" rules are left to the specific control plane API generators. 

From the control plane side of things we should say that the type substitution should not performed, i.e. if an object type has been declared with a certain type, that information (including annotations) must be preserved and made available to the API generators. 

I would not try to state where the API generation must be happening or which optimizations should or should not take place before it -- different people might have different approaches there and in some cases these might be user-controlled options. For example, some people might choose to specifically include automatically eliminated tables in the API for the purposes of backwards/forwards compatibility of their control plane code, and some would definitely not want that.

**Comment by jnfoster** [comment]
This discussion is fascinating, but seems to have stalled. Speak up if we should keep it open as an active issue.
